main.go

```diff
@@ -241,22 +275,50 @@ func main() {
 			log.Println(err.Error())
 		}
 
-		point := influxdb2.NewPoint(
-			*measurementName,
-			map[string]string{"ups_name": *upsNameTag}, // tags
-			fields,
-			atTime,
-		)
-		if err := retry.Do(
-			func() error {
-				ctx, cancel := context.WithTimeout(context.Background(), influxTimeout)
-				defer cancel()
-				return influxWriteAPI.WritePoint(ctx, point)
-			},
-			retry.Attempts(2),
-		); err != nil {
-			log.Printf("failed to write point to influx: %v", err)
-		} else if hb != nil {
+		var influxSuccess bool
+		if influxWriteAPI != nil {
+			point := influxdb2.NewPoint(
+				*measurementName,
+				map[string]string{"ups_name": *upsNameTag}, // tags
+				fields,
+				atTime,
+			)
+			if err := retry.Do(
+				func() error {
+					ctx, cancel := context.WithTimeout(context.Background(), influxTimeout)
+					defer cancel()
+					return influxWriteAPI.WritePoint(ctx, point)
+				},
+				retry.Attempts(2),
+			); err != nil {
+				log.Printf("failed to write point to influx: %v", err)
+			} else {
+				influxSuccess = true
+			}
+		}
+
+		var mqttSuccess bool
+		if mqttClient != nil && mqttClient.IsConnected() {
+			mqttSuccess = true
+			for fieldName, fieldValue := range fields {
+				topic := fmt.Sprintf("%s/%s", *mqttTopic, fieldName)
+				payload := fmt.Sprintf("%v", fieldValue)
+
+				if err := retry.Do(
+					func() error {
+						token := mqttClient.Publish(topic, 0, false, payload)
+						token.Wait()
+						return token.Error()
+					},
+					retry.Attempts(2),
+				); err != nil {
+					log.Printf("failed to publish %s to MQTT: %v", fieldName, err)
+					mqttSuccess = false
+				}
+			}
+		}
+
+		if hb != nil && (influxSuccess || mqttSuccess) {
```

Comment:
What if Influx and MQTT are both configured? This logic should only update the heartbeat timestamp if **all** configured outputs succeeded.